### PR TITLE
Syncronize access to `func performSync(_ block: @escaping (Realm) -> Void) {`, to make sure there is no two or more call at same time #4883

### LINKED ISSUE #4883: Fix crash
*** -[AlphaWallet.RunLoopThread performSelector:onThread:withObject:waitUntilDone:modes:]: target thread exited while waiting for the perform

### PR DESCRIPTION
**Pull request by oa-s**
Fixes #4883

### CODE REVIEW

File: AlphaWallet/Core/Types/RealmStore.swift [hunk 1]
```diff
@@ -31,12 +32,18 @@ class RealmStore {
...
+    func performSync(_ block: @escaping (Realm) -> Void) {
         if Thread.isMainThread {
-            callback(mainThreadRealm)
+            block(mainThreadRealm)
         } else {
-            thread.performSync() {
-                callback(self.backgroundThreadRealm)
+            //NOTE: synchronize calls from different threads to avoid
+            //*** -[AlphaWallet.RunLoopThread performSelector:onThread:withObject:waitUntilDone:modes:]: target thread exited while waiting for the perform
+            dispatchPrecondition(condition: .notOnQueue(queueForRealmStore))
```
Comment (hboon) [inline AlphaWallet/Core/Types/RealmStore.swift:41]: This line crashes if it's not running on `queueForRealmStore`. Where do we enforce that `performSync(:)` (i.e. its caller) always runs on either `.main` or `queueForRealmStore `?

File: AlphaWallet/Core/Types/RealmStore.swift [hunk 2]
```diff
@@ -31,12 +32,18 @@ class RealmStore {
         thread.start()
     }
 
-    func performSync(_ callback: @escaping (Realm) -> Void) {
+    func performSync(_ block: @escaping (Realm) -> Void) {
```
Comment (hboon) [inline AlphaWallet/Core/Types/RealmStore.swift:35]: I don't understand the original `RealmStore.performSync(_:)` before the PR. Does it work like that:

```
func performSync(block)
    if currentThread == .mainThread
        run(block, .mainThread + mainThreadRealm)
    else
        run(block, .backgroundThread + backgroundThreadRealm)
```

Why is it optionally switching between 2 realm objects based on the caller of `performSync(:)`? Why not just one, say on .main (or background) and then always go to it?

File: AlphaWallet/Core/Types/RunLoopThread.swift [hunk 3]
```diff
@@ -43,14 +43,14 @@ class RunLoopThread: Thread {
         }
     }
 
-    func performSync(_ block: @escaping () -> Swift.Void) {
```
Comment (oa-s) [inline AlphaWallet/Core/Types/RunLoopThread.swift:46]: `performSync` - get called for all calls related to realm, when we fetch or write data to realm, all operations call on run loop thread. (maybe function name isn't good)

File: AlphaWallet/Core/Types/RealmStore.swift (hunk 2 above)
Comment (oa-s) [reply AlphaWallet/Core/Types/RealmStore.swift:35]: > Does it work like that:

yes, pretty like you described

File: AlphaWallet/Core/Types/RealmStore.swift (hunk 1 above)
Comment (oa-s) [reply AlphaWallet/Core/Types/RealmStore.swift:41]: the `else`  part could be called on any non main thread,  and we use 
```           
dispatchPrecondition(condition: .notOnQueue(queueForRealmStore))
queueForRealmStore.sync {
```
to make sure the no one other is trying to call in same time

**Review by hboon — APPROVED** [review]

**Pull request merged**